feat: Add custom states

PythonGermany opened the pull request:
<!-- Thank you for contributing! -->

## Summary
<!-- If there's a relevant issue, you can just write the issue number below (e.g. #123) -->
Implement the requested severity feature by introducing custom states.

If implementation moves forward as  suggested:
Closes #227, close #1274.

If we structure it correctly we could start working on #1444 as a followup if this PR gets merged to make it possible to configure multiple themes for each state to allow adding colorblind  and other themes which can then be selected via the UI.

## Checklist
<!-- Replace [ ] by [X] if you have completed the item -->
### Implementation tasks:
The following logic needs to be implemented (backend):
- [x] New config section `states` defining the available states
  - [x] No custom config is given -> Set default states `healthy`, `unhealthy` and `maintenance`
  - [x] Merge custom config with default config
  - [x] Validate config
    - [x] Check name conflicts
    - [x] Check priority conflicts
- [x] New UI section configuring colors/themes for states
  - [x] No custom config is given -> Set default color config for default states
  - [x] Merge custom config with default config
  - [x] Validate config
    - [x] Verify that names in state and color config match
    - [x] Validate color code format
 - [ ] Extend alerting config
   - [ ] Add state priority threshold below which no alerting should take place (set state in config and resolve to priority at runtime)
- [x] Logic to evaluate state of result/endpoint
  - [x] In maintenance state
  - [x] Default healthy/unhealthy states -> One properly implemented also covered by severity logic
  - [x] Custom states -> Input will be condition evaluation results
    - [x] Select correct state based on highest priority
- [x] New API functionality
  - [x] Send color and UI relevant information for states
  - [x] Send new state information of result
  - [x] Include state in events to display useful info and correct color in UI
  - [ ] Handle custom states for external endpoints
    - [x] Handle healthy and unhealthy defaults
    - [ ] Allow for custom state input
      - [ ] Validate custom state against current config? If not valid default to healthy unhealty?
    - [ ] Handle maintenance state 
- [ ] Handle custom states for suites
- [x] DB Migration adding new result state column if not present and set appropriate state based on Success value

The following logic has to be implemented (frontend):
- [x] Add new placeholders to embed color config when starting gatus
- [ ] Handle current status UI text value in endpoint details view
- [x] Use new API state fields
  - [x] to display correct color and text in status badges
  - [x] to display correct color and text in endpoint events
  - [x] to display correct color and text in endpoint result history bar
  - [x] to display correct color and text for events in response time trend chart
- [x] Handle absence of custom color config (to prevent issues in development environment or state config changes)
  - Handled by using a hardcoded frontend local color used to communicate missing custom state color config 
### Outstanding issues/tasks/questions

- [x] Colors are broken when in frontend development mode -> `npm run serve`
- [x] Add new event on state change and not just on success change?
  - [ ] Make configurable?
- [ ] Test impacts of color stuff on remote instance experimental feature
- [ ] Reserve `unknown` state and allow state color config?
- [ ] Show state for result in tooltip
- [ ] Set success field based on new config option configurable for each state
- [ ] Apply custom color for default states to everything relevant regarding suites
- [x] How is config file merging handled? First merged then validated? (interest because of new config sections)
  - readme says first merged then validated

### Final tasks
- [ ] Tested and/or added tests to validate that the changes work as intended, if applicable.
- [ ] Updated documentation in `README.md`, if applicable.

PythonGermany committed:
refactor(ui): Cleanup template option handling

PythonGermany committed:
chore: Add state configuration prototype

PythonGermany committed:
refactor: Remove currently unused field

PythonGermany committed:
chore: Add hardcoded maintenance color

PythonGermany commented:
Thanks for the heads up! I understand  that working on this might lead nowhere and that it's a complicated thing to add since it impacts a lot of core components. Currently I'm just exploring and trying to figure out what the biggest road blocks are towards a fully working implementation/prototype.

Out of my perspective it's a good topic because this way I need  to have a look at many different core components and features of Gatus. So even if it leads nowhere it will not be wasted time for me, I'm sure.

> The issue is usually that it's a breaking change

Breaking change in configuration compatibility is probably the most critical I imagine. How problematic are breaking changes in API and database structure (e.g. added state member, in addition to the currently existing success field)?

> I'm not even convinced I would want this feature in Gatus in the first place because of all the confusion it adds around health % calculation, uptime calculation

My goal would be to not change the default behavior in any way. Only people trying to would have to think about the implications of customizing it.

In regards to added complexity I definitely see your point. However I think is the structure on how to configure it is well designed one detailed entry in the FAQ section of the readme would be enough for most people to understand.

I'll keep this PR in draft as long as I'm experimenting and once the base of the implementation is solid and thorough enough I'll open it. Maybe you'll see the potential once I have hopefully solved some of the bigger issues, and if not so be it. I know what I signed up for!

TwiN commented:
Making a breaking change on the database schema would only imply removing or modifying existing fields. It wouldn't be preferable, but it's not a big deal if we have to do that. There is no public database API contract for Gatus, meaning there is no guarantee that the database schema will not have breaking changes.

The REST API, however, is a different story.
In your case, you'd likely be adding to the API, not removing or changing existing fields in a way that would actually constitute a breaking change.

Changes are just breaking if users currently relying on a specific field are no longer able to rely on that field due to a change. Such a change would be breaking, and would require a major version bump, which can be done if absolutely necessary.

If you're not sure whether something is a breaking change, feel free to ping me and ask.

PythonGermany committed:
chore: Add maintenance UI showcase

PythonGermany committed:
chore: Add backend state color configuration

PythonGermany committed:
chore: Use backend color config in UI

PythonGermany committed:
chore: Test darker color for better visibility

PythonGermany committed:
cleanup: Remove unused prototype code

PythonGermany committed:
test: Update config for testing

PythonGermany committed:
chore: Allow linking condition to states

PythonGermany committed:
chore: Store state for persistent storage types

PythonGermany committed:
chore: Use lighter unhealthy color

PythonGermany committed:
feat: Evaluate state based on priority

PythonGermany committed:
chore: Validate correct hex code format

PythonGermany committed:
fix: Restore original maintenance log order

PythonGermany committed:
fix: Broken colors in development mode

PythonGermany committed:
feat: Provide state for each event

PythonGermany committed:
feat(ui): Use event state in endpoint details view

PythonGermany committed:
chore(ui): Use utils color function

PythonGermany committed:
refactor: Move badge color logic to StatusBadge

PythonGermany committed:
fix(ui): Toolip result highlighting

PythonGermany committed:
chore: Use lighter unhealthy color

PythonGermany committed:
refactor: Only validate user input colors

Default colors should be validated in unit tests

PythonGermany committed:
refactor: Create type for color

Can be used later if validation for type is extended and in unit tests
to valiate default config once during testing and not always at runtime

PythonGermany committed:
cleanup(ui): Remove unused component

PythonGermany committed:
refactor: Update event type structure and init

PythonGermany committed:
fix: Unknown state label handling

PythonGermany committed:
fix: Add missing event query column

PythonGermany committed:
fix: Incorrect default state name

PythonGermany committed:
fix: Improve event messages

PythonGermany committed:
Merge remote-tracking branch 'origin/master' into add-custom-state-support

PythonGermany committed:
fix: Add missing local variable

PythonGermany committed:
test: Add for linked condition states

PythonGermany committed:
refactor: Change error message

PythonGermany committed:
test: Add for state config

PythonGermany committed:
cleanup: Remove resolved todos

PythonGermany committed:
chore: Update custom states test config

PythonGermany committed:
chore(ui): Regenerate static assets

PythonGermany committed:
refactor: Get color directly

PythonGermany committed:
fix: Create events when state has changed

PythonGermany committed:
feat(ui): Apply custom states to response time char

PythonGermany committed:
chore(ui): Regenerate static assets

PythonGermany committed:
test: Update suite config for testing

PythonGermany committed:
fix(storage): Add new column migration

PythonGermany committed:
cleanup: Omit empty states from json

PythonGermany committed:
chore: Check valid name characters

PythonGermany committed:
fix(ui): Use nodata color for badge if with no result

PythonGermany committed:
refactor: Rename to match new state system

PythonGermany committed:
cleanup: Remove resolved storage TODOs

PythonGermany committed:
refactor(ui): Make local state color names consistent

PythonGermany committed:
fix: No default priority linked for confitions

PythonGermany committed:
Merge remote-tracking branch 'TwiN/master' into add-custom-state-support

PythonGermany committed:
Merge branch 'master' into add-custom-state-support

PythonGermany committed:
fix: Improve condition validity checks

PythonGermany committed:
Merge remote-tracking branch 'TwiN/master' into add-custom-state-support

PythonGermany committed:
fix: Add missing config field removed in merge

PythonGermany committed:
feat: External endpoint custom state handling

PythonGermany committed:
chore(ui): Regenerate static assets

PythonGermany committed:
chore: Update custom state test config

PythonGermany committed:
Merge remote-tracking branch 'TwiN/master' into add-custom-state-support

PythonGermany committed:
fix: Substitute state in color method

PythonGermany committed:
fix: External endpoint timeout result state

PythonGermany committed:
chore(ui): Regenerate static assets

Zophar78 commented:
I've tested your branch and it works as intended for me (And i'm really looking forward to add a "degrated" mode for my use case which is a requirement).
Just my two-cents on the configuration: it would be "easier" to have the cutom state color directly in the "states" section like this:
```bash
states:
  - name: degraded
    priority: 1
    color: "#FFA500"  # Orange color for degraded state
```

instead of:
```bash
states:
  - name: degraded
    priority: 1

ui:
  state-colors:
    degraded: "#FFA500"  # Orange color for degraded state
```

It would replacate states in ui section as well.

marcaurele commented:
> Just FYI before you commit time to implementing this - there has been a few attempts at implementing severity, and the issue is usually that it's a breaking change and I don't like the implementation. I understand some people would love a way to add some level of severity, but to me, I'm not even convinced I would want this feature in Gatus in the first place because of all the confusion it adds around health % calculation, uptime calculation, the increase in complexity for specifying what severity a specific condition impacts, etc.
> 
> I've seen the comment you posted on the issue, and it's a good idea to avoid breaking changes, but I think it would be too difficult to understand for most people (and also `states` is perhaps not the right word here for this configuration, as it's not really self-explanatory like `severities` or something similar would be), and again, it would add a lot of complexity for people to understand in the configuration.
> 
> Long story short, it adds a whole lot of complexity for what I think is very little advantage, and for that reason, at this stage, I'm not sure I'm interested in adding this to Gatus

One interesting case for the degraded status is the TLS certificate expiry. It's more granular to be able to turn an endpoint as degraded if the certificate will expire in the next 7 days, and turning to unhealthy when it's expired. _Unhealthy_ is too strong in that case as it's a warning and not a service downtime state.

christophdb commented:
**A concrete production use case for custom states — with a note on the uptime concern**

We monitor SeaTable Cloud with Gatus on a customer-facing status page. Our backend runs **multiple dtable-servers**, but we deliberately *don't* want one Gatus endpoint per server — that's too technical and confusing for customers. Instead we expose a single **"meta" health endpoint** on our side that internally checks all dtable-servers and rolls them up into one status.

What we're missing is exactly a **third state**:
- 🟢 healthy — all servers fine
- 🟡 degraded — at least one server has issues, but the service is still reachable/usable
- 🔴 down — everything is unavailable

**Where the logic lives:** the actual aggregation ("which of our N servers is unhealthy?") stays entirely on our side, inside that meta endpoint. Gatus polls it like any normal endpoint and only needs to map the response to a state. Our endpoint always returns `200` with a small JSON body, e.g.:
```json
{ "status": "degraded", "affected": ["dtable-3"] }
```
and Gatus would select the state via a simple condition such as `[BODY].status == healthy` / `== degraded` / `== down`. So Gatus isn't computing severity across many conditions — it's a trivial status→state mapping on a single endpoint, using the same condition mechanism it already has. The complexity @TwiN was worried about doesn't enter the picture here.

**On the uptime question (the main worry raised):** in our case `degraded` must count as `up` for the uptime/availability %. Our SLA is "reachable/usable", and the service *is* available when only one backend server is degraded — the yellow state communicates the partial issue visually without distorting the availability metric. That said, I can easily imagine other operators wanting degraded to count as downtime, which is exactly why the **per-state `success` mapping** already on your checklist (configurable whether a state counts toward uptime) is the right design — it lets each operator pick the semantics instead of Gatus imposing one.

This is the same spirit as @marcaurele's TLS-expiry example: a "warning, not an outage" state that's genuinely useful and shouldn't be forced into the binary up/down model. Happy to help test this against a real multi-server setup if useful.

christophdb commented:
I tested this feature against the current master and it works great. 🎉

Since the branch was ~4 months behind, I merged the latest master into it first. The conflicts were all trivial (master's new LoginSubtitle vs. this branch's StateColors, plus regenerated frontend assets), none in the custom-states logic. After that: go build ./... ✅, the ui / state / sql tests pass ✅, and the three states render correctly end-to-end (the API returns state: degraded and the custom color shows up).

Here's a real-world example of why this matters for us: we run SeaTable Cloud with multiple dtable-servers behind a single "meta" endpoint. We don't want one endpoint per server (too technical for customers). We just want 🟢 healthy / 🟡 degraded (≥1 server has issues but the service is still usable) / 🔴 down. This feature does exactly that:

<img width="1285" height="596" alt="Screenshot_2026-06-12_11-03-51" src="https://github.com/user-attachments/assets/2f402b2c-5c7a-4c1e-8cd4-acd1a8cf6d90" />

I pushed the rebased/merged branch as a PR into your fork so it's easy to pull in if useful: PythonGermany/gatus#26. No pressure at all, just trying to help move this forward. Really hope this lands; thanks for the excellent work! 🙏

christophdb commented:
Sorry for the next comment... but two more thoughts:

### 1) another use case:
I've been playing around with this some more and honestly it's fantastic. 🙌

I found another use case that I think is even more universal than the multi-backend one I mentioned earlier: using a custom state for slow response times. "Slow but reachable" really shouldn't show up as a full outage, and with this feature it just works:

```
conditions:
  - "[STATUS] == 200"                  # no/error response -> unhealthy (red)
  - "degraded::[RESPONSE_TIME] < 100"  # slow -> degraded (yellow); else healthy
``` 
Here it is running on a endpoint (a login page). The result-history bar nicely shows the degraded buckets where the response time crossed the threshold, with the condition breakdown right there in the tooltip:

<img width="1397" height="385" alt="Screenshot_2026-06-12_13-06-29" src="https://github.com/user-attachments/assets/2889996a-8106-4889-89e0-acb102d1dbef" />

### 2) alerting + uptime
One more thing I ran into while testing. The interaction between custom states and alerting + uptime, which I think is the key question for real-world use.

Right now a `degraded::` condition that fails sets `result.Success = false` (same as any failing condition), and both alerting and the uptime % key off that boolean. So as it stands, a degraded result triggers notifications just like a full outage (counts toward `failure-threshold`) and counts as downtime in the uptime %. I see this is already on your radar (the TODO#227 in `watchdog/endpoint.go`) and the two unchecked alerting items in the checklist.

What makes this interesting is that the same degraded state needs opposite alerting behavior depending on the check. Two real examples from our setup:

1. Slow response time (`degraded::[RESPONSE_TIME] < 1000`) should show 🟡 on the dashboard but not page anyone. It's usually transient, "slow but reachable" isn't an emergency, and it shouldn't drag down the availability SLA.
2. TLS cert expiring in <7 days (`degraded::[CERTIFICATE_EXPIRATION] > 168h`) should show 🟡 and absolutely send a notification, because that's exactly the early warning you want in order to renew in time. Here a missed alert is the whole problem.

So a single global "degraded doesn't alert" rule wouldn't work for either of us. Case 1 wants silence, case 2 wants a page, both on the same state. That's what makes me think an alertable switch per check (alongside a per-state up/down mapping for the uptime %) would be the flexible primitive here.

Is that roughly the direction you're considering for the success/priority config? Happy to help test that part once it exists and thanks again, the visual side already works great.

breart commented:
Are there any updates? This functionality would be extremely valuable. 

How we can move this forward? @TwiN

K-e-l-v-i-n-12 commented:
Unfortunately, we can't use Gatus because it doesn't distinguish between maintenance and outages. We're eagerly awaiting this feature.
@TwiN : Would it be possible to get a quick response as to whether this is planned and whether it's worth waiting, or if we'll have to rely on other heavyweights after all—which we'd really rather not do?
